File: CSW/C++/220907_최소원_컴백홈.cpp
#include <iostream>
#include <string>
using namespace std;

int map[5][5];
int visited[5][5];
int R, C, K, cnt;
string s;

int dirY[4] = { 1, 0, -1, 0 };
int dirX[4] = { 0,1,0,-1 };


void dfs(int y,int x, int dis) {
	if (y == 0 && x == C - 1 && dis == K) {
		cnt++;
		return;
	}

	for (int i = 0; i < 4; i++) {
		int dy = y + dirY[i];
		int dx = x + dirX[i];
		if (dy < 0 || dx < 0 || dy >= R || dx >= C) continue;
		if (visited[dy][dx] || map[dy][dx]) continue;
		visited[dy][dx] = 1;
		dfs(dy, dx, dis + 1);
		visited[dy][dx] = 0;
	}
}

int main() {
	
	cin >> R >> C >> K;

	for (int i = 0; i < R; i++) {
		cin >> s;
		for (int j = 0; j < C; j++) {
			if (s[j] == 'T') map[i][j] = 1;
		}
	}


	visited[R - 1][0] = 1;
	dfs(R - 1, 0, 1);

	cout << cnt;

	return 0;
}

File: CSW/JavaScript/220907_최소원_컴백홈.js
var fs = require('fs');
var input = fs.readFileSync(`/dev/stdin`).toString().trim().split('\n');

let [R, C, dist] = input[0].split(' ').map(e=>parseInt(e));

const map =Array.from({length:R}, ()=>Array(C).fill(0));
const visited =Array.from({length:R}, ()=>Array(C).fill(0));

for(let i=1; i<=R; i++){
    for(let j=0; j<C; j++){
        if(input[i][j]==='T') map[i-1][j] = 1;
    }
}

const dirY = [1,0,-1,0];
const dirX = [0,1,0,-1];
let cnt = 0;

const dfs = (y, x, dis) =>{
    if(y===0 && x===C-1 && dis ===dist ){
        cnt++;
        return;
    }
    for(let i=0; i<4; i++){
        let dy = y + dirY[i];
        let dx = x + dirX[i];
        if(dy<0 || dx<0|| dy>=R || dx>=C)continue;
        if(visited[dy][dx] || map[dy][dx]) continue;
        visited[dy][dx] = 1;
        dfs(dy,dx, dis+1);
        visited[dy][dx] = 0;
    }
}

visited[R-1][0] =1;
dfs(R-1, 0, 1);


console.log(cnt);

